### ISSUE TITLE
Advice on how to tune and debug Horovod performance

### ISSUE THREAD

**Issue by abhishekpatnia** [uber/horovod#817]
Hi, 
I ran synthetic CNN benchmark on a 4 GPU node(1080 Ti). I got results that seemed reasonable. 

<img width="458" alt="screen shot 2019-02-07 at 7 21 17 pm" src="https://user-images.githubusercontent.com/6693858/52457225-92dc6f80-2b0d-11e9-906d-a80de82abe48.png">

Then, I tried Horovod on a simple dummy model. Below is the code.

```
import time

import horovod.tensorflow as hvd
import numpy as np
import tensorflow as tf
from tensorflow.python.keras import backend as K

tf.logging.set_verbosity(tf.logging.INFO)

hvd.init()
tf_config = tf.ConfigProto()
tf_config.gpu_options.allow_growth = True
tf_config.gpu_options.visible_device_list = str(hvd.local_rank())


def get_model(projection_dims, batch_size):
    x_input = tf.keras.layers.Input(batch_shape=(batch_size, projection_dims), dtype=tf.float32)
    x = tf.keras.layers.Dense(projection_dims, activation="relu")(x_input)
    x = tf.keras.layers.Dense(projection_dims, activation="relu")(x)
    x = tf.keras.layers.Dense(projection_dims, activation="relu")(x)
    x = tf.keras.layers.Dense(projection_dims, activation="relu")(x)
    x = tf.keras.layers.Dense(5000)(x)

    model = tf.keras.Model(inputs=x_input, outputs=x)

    return model


def train():
    batch_size = 512
    projection_dim = 5120
    K.set_learning_phase(1)
    Y_true = tf.placeholder(tf.int32, name="Y", shape=(batch_size,))

    model = get_model(projection_dim, batch_size)

    batch_loss = tf.losses.sparse_softmax_cross_entropy(Y_true, model.output)
    optimizer = tf.train.RMSPropOptimizer(learning_rate=0.2 * hvd.size())
    optimizer = hvd.DistributedOptimizer(optimizer, compression=hvd.compression.FP16Compressor)
    gradients, variables = zip(*optimizer.compute_gradients(batch_loss))
    train_op = optimizer.apply_gradients(zip(gradients, variables))

    bcast = hvd.broadcast_global_variables(0)
    init_op = tf.global_variables_initializer()

    with tf.Session(config=tf_config) as sess:
        init_op.run()
        bcast.run()
        tima_a = time.time()
        X = np.ones((batch_size, projection_dim), dtype=np.float32)
        Y = np.ones((batch_size,), dtype=np.float32)
        for i in range(1, 501):
            _, batch_loss_val = sess.run([train_op, batch_loss], feed_dict={model.inputs[0]: X,
                                                                            Y_true: Y})
            if i % 100 == 0:
                duration = time.time() - tima_a
                tf.logging.info("Worker: {0} Time Taken(s): {1}".format(hvd.rank(), duration))
                tima_a = time.time()


train()
```

First I tried this with a single GPU and below are the times for 100 batches that I print for first 500 batches.

<img width="416" alt="screen shot 2019-02-07 at 7 39 52 pm" src="https://user-images.githubusercontent.com/6693858/52457716-29aa2b80-2b10-11e9-8a74-7a406208d2b9.png">

Then, I tried with 2 GPUs using the following command

```
mpirun -np 2 -H localhost:2 -bind-to none -map-by slot -x NCCL_DEBUG=INFO  -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python model.py 
```

But, now I don't see the advantage of using 2 GPUs. Can you educate me on what am I doing wrong here?

<img width="425" alt="screen shot 2019-02-07 at 7 46 07 pm" src="https://user-images.githubusercontent.com/6693858/52457897-0633b080-2b11-11e9-9b55-a701f7f76882.png">

**Comment by abhishekpatnia** [comment]
I am closing this issue. I read more closed issues here and also played around with the configuration. I realize now that we need to saturate the per worker load before switching it to multi-gpu setup. This would make per-worker computation bigger than the inter-GPU communication and then we will see speed-up. 

Please reopen it if you want to add more.
